### ISSUE TITLE
Additive KISS-GP (SV-DKL) outputs constant mean after training

### ISSUE THREAD

**Issue by Akella17** [cornellius-gp/gpytorch#557]
AbstractVariationalGP module, AdditiveGridInterpolationVariationalStrategy.

Please find the ipython notebook in this link: https://drive.google.com/file/d/1Dg1bEb3_5g9Rf-BIBT71kyj4Q2aMy9p0/view?usp=sharing

I observed that the model's mean function learns a constant value, i.e. the mean starts from 0 at the beginning of training and converges to the average of the training outputs. When it is able to learn the mean value of training outputs, why is it not able to capture its entire structure.

While this is not relevant to the abovementioned issue, I noticed from my previous experiments with working SV-DKL models that the network captures the complexity of training outputs in a gradual manner. Assuming that the training outputs are of the form quadratic + sinusoid (like in the ipython notebook), the network's mean function remains constant for a few iterations while it tries to change its value to minimize the loss. Then the network's mean function tries to approximately model the parabola. Only after training through sufficient epochs does the network learn to capture the variations of the sinusoid. I just want to know if there is a reason for observing this kind of gradual behavior as opposed to the random contours throughout the training of plain neural networks?

**Comment by jacobrgardner** [comment]
Hi @Akella17 ,

I ran your code and ran in to the same problem with several different variational strategies. This leads me to believe that something else strange is going on with the model, which I'm not sure I fully understood (e.g., the train_x you are plotting against is 4D, but the inputs to the GP end up being 64D). 

If you think there's a bug do you maybe have simpler version that reproduces? Alternatively, If you're just looking for modelling advice, do you think you could maybe supply a little bit more information about what's going on here?

**Comment by Akella17** [comment]
Oh, in an attempt to simplify the code I might have forgot to change 64 to 4. However, it does not seem to have affected the training. The mean function still remains constant. And to make sure that the reason for this behavior is not insufficient training, I have also reduced the grid_size from 256 to 64 (which I believe in some sense restricts the expressive power of the model; correct me if I am wrong in assuming so :p). In the updated notebook, both train_x and GP input_dims are 4D.

Please have a look at the updated notebook: https://drive.google.com/file/d/1Dg1bEb3_5g9Rf-BIBT71kyj4Q2aMy9p0/view?usp=sharing

We are trying to implement an SV-DKL model in reinforcement learning domain. The code that I have shared is a simplification of the actual network while capturing the bug I am encountering. Moreover the code is very similar to the *Large-Scale Stochastic Variational GP Regression (CUDA) (w/ KISS-GP)* example notebook (https://gpytorch.readthedocs.io/en/latest/examples/05_Scalable_GP_Regression_Multidimensional/SVDKL_Regression_GridInterp_CUDA.html).

In case you are running the code locally, you can skip the first 3 code snippets, which essentially set up the environment in Colabs.

**Comment by Akella17** [comment]
I think I have isolated the source of this issue. For robust training of SV-DKL models, I believe that we shall be needing different sets of learning rates for the neural network and the GP layers. For robust training, I observed that GP layers require LR in the range (1e-2, 1e-1) whereas neural networks require LR in the range (1e-3, 1e-4).

Also, the choice of learning rate is determined by the activation function in use, i.e. linear layers and ReLU activations are robust to high LR with gradient clipping but Tanh and Sigmoid activations require LR in the range (1e-3, 1e-4).

**Comment by jacobrgardner** [comment]
@Akella17 That makes some sense. I believe in the CIFAR10/100 example and in a few other DKL examples we use different learning rates for the neural network and the GP, as well.

Adding fancy modeling choices like neural networks etc in general will require more tuning to get the value out of them.
